src/D3MOracle.sol

```diff
@@ -0,0 +1,117 @@
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+pragma solidity ^0.8.14;
+
+interface VatLike {
+    function live() external view returns (uint256);
+}
+
+interface HubLike {
+    function culled(bytes32) external view returns (uint256);
+}
+
+contract D3MOracle {
+    // --- Auth ---
+    /**
+        @notice Maps address that have permission in the Pool.
+        @dev 1 = allowed, 0 = no permission
+        @return authorization 1 or 0
+    */
+    mapping (address => uint256) public wards;
+    address public hub;
+
+    VatLike public immutable vat;
+    bytes32 public immutable ilk;
+
+    uint256 internal constant WAD = 10 ** 18;
+
+    // --- Events ---
+    event Rely(address indexed usr);
+    event Deny(address indexed usr);
+    event File(bytes32 indexed what, address data);
+
+    constructor(address vat_, bytes32 ilk_) {
+        vat = VatLike(vat_);
+        ilk = ilk_;
+
+        wards[msg.sender] = 1;
+        emit Rely(msg.sender);
+    }
+
+    /// @notice Modifier will revoke if msg.sender is not authorized.
+    modifier auth {
+        require(wards[msg.sender] == 1, "D3MOracle/not-authorized");
+        _;
+    }
+
+    // --- Administration ---
+    /**
+        @notice Makes an address authorized to perform auth'ed functions.
+        @dev msg.sender must be authorized.
+        @param usr address to be authorized
+    */
+    function rely(address usr) external auth {
+        wards[usr] = 1;
+        emit Rely(usr);
+    }
+
+    /**
+        @notice De-authorizes an address from performing auth'ed functions.
+        @dev msg.sender must be authorized.
+        @param usr address to be de-authorized
+    */
+    function deny(address usr) external auth {
+        wards[usr] = 0;
+        emit Deny(usr);
+    }
+
+    /**
+        @notice update vow or end addresses.
+        @dev msg.sender must be authorized.
+        @param what name of what we are updating bytes32("vow"|"end")
+        @param data address we are setting it to
+    */
+    function file(bytes32 what, address data) external auth {
+        require(vat.live() == 1, "D3MOracle/no-file-during-shutdown");
+
+        if (what == "hub") hub = data;
+        else revert("D3MOracle/file-unrecognized-param");
+        emit File(what, data);
+    }
+
+    /**
+        @notice Return value and status of the oracle
+        @return val always 1 WAD
+        @return ok true if vat is live or ilk is not culled
+    */
+    function peek() public view returns (uint256 val, bool ok) {
+        val = WAD;
+        ok = vat.live() == 1 || HubLike(hub).culled(ilk) == 0;
```

Comment:
Subjective (and we might want to ignore at this stage) but an alternative of having the logic for `ok` here is calling a new view function in the hub (`hub.endable(ilk)`?) and having the logic there.

The conditions for `ok` are really entwined with the hub states, so this will support updating the hub without updating this oracle. Will also save us from having `vat` here and add some clarity for this condition through the function name (though for that we can just add a comment here).

Reply:
We still need `vat` for the `file` function so I'd say we keep as it is.
If we update the `hub` we still need to update the address in the oracles so the only difference would be we need to make sure to not break anything of the structure and being everything in the same repo makes it quite resistant to that kind of errors.

Reply:
Ok. Maybe only add a comment of how this relates to the End?
Just to remove the required knowledge for understanding why the condition was added.